### PR TITLE
namespace: remove gc finalizers based on delete options

### LINKED ISSUE #76042: Namespace storage does not remove OrphanFinalizer during cascade deletion
**What happened**:

I created a namespace `foo` with `orphan` finalizer, and then tried to delete it:

```console
$ cat ns.yaml
apiVersion: v1
kind: Namespace
metadata:
  name: foo
  finalizers:
  - orphan
$ kubectl create -f ns.yaml
namespace/foo edited
$ kubectl delete ns foo
namespace "foo" deleted
```

The namespace can be deleted, and the `orphan` finalizer is removed by garbage collection controller:

```
...
I0401 19:09:56.510632   77981 graph_builder.go:578] GraphBuilder process object: v1/Namespace, namespace , name foo, uid c396b742-9be5-4d5b-b86a-4614853e9516, event type update
I0401 19:09:56.510672   77981 graph_builder.go:540] add [v1/Namespace, namespace: , name: foo, uid: c396b742-9be5-4d5b-b86a-4614853e9516] to the attemptToOrphan
I0401 19:09:56.510744   77981 garbagecollector.go:570] successfully updated all dependents of owner [v1/Namespace, namespace: , name: foo, uid: c396b742-9be5-4d5b-b86a-4614853e9516]
```

**What you expected to happen**:

Since kubectl species `propagationPolicy: Background`, it is not an orphaning operation, the finalizer should be removed in namespace store.

**How to reproduce it (as minimally and precisely as possible)**:

Same as what I observed:

1. create a namespace.
1. add `orphan` finalizer.
1. run `kubectl delete`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): HEAD

### PR DESCRIPTION
**Pull request by zhouhaibing089**
This makes the behavior being consistent with generic store, The
orphan finalizer should be removed if the delete options does not
specify propagarionPolicy as orphan.

/kind bug
/sig api-machinery

Fixes https://github.com/kubernetes/kubernetes/issues/76042

```release-note
NONE
```

### CODE REVIEW

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 1]
```diff
@@ -189,6 +189,91 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
+				Finalizers: []api.FinalizerName{},
+			},
+			Status: api.NamespaceStatus{Phase: api.NamespaceActive},
+		}
+		if err := storage.store.Storage.Create(ctx, key, namespace, nil, 0, false); err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		if obj, _, err := storage.Delete(ctx, test.name, test.deleteOptions); err != nil {
+			t.Errorf("unexpected error: %v", err)
+		} else {
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: Use t.Fatal in the first case and remove the "else" here to reduce indention.

File: pkg/registry/core/namespace/storage/storage.go [hunk 2]
```diff
@@ -185,18 +185,27 @@ func (r *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOp
...
+					(options.PropagationPolicy != nil && *options.PropagationPolicy == metav1.DeletePropagationOrphan) {
+					newFinalizers = append(newFinalizers, metav1.FinalizerOrphanDependents)
+				}
+
+				// add foregroundDeletion if the options says foreground deletion.
+				if options.PropagationPolicy != nil && *options.PropagationPolicy == metav1.DeletePropagationForeground {
+					newFinalizers = append(newFinalizers, metav1.FinalizerDeleteDependents)
+				}
+
+				existingNamespace.ObjectMeta.Finalizers = newFinalizers
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage.go]: Check if finalizers have changed to avoid unnecessary update. https://github.com/kubernetes/kubernetes/blob/20cece67b5815f6e37ee72b86f282f027bab820b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L740-L744

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 3]
```diff
@@ -189,6 +189,91 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
+	for _, test := range tests {
+		key := "namespaces/" + test.name
+		ctx := genericapirequest.NewContext()
+		namespace := &api.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: test.name,
+				Finalizers: []string{
+					metav1.FinalizerOrphanDependents,
+					metav1.FinalizerDeleteDependents,
+				},
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: Also test a namespace that originally doesn't have finalizers?

File: pkg/registry/core/namespace/storage/storage.go (hunk 2 above)
Comment (caesarxuchao) [reply pkg/registry/core/namespace/storage/storage.go]: Also this is not the logic of the generic store, which will preserve the existing finalizers in the object is the deleteOptions doesn't explicitly set the mode of the deletion. See https://github.com/kubernetes/kubernetes/blob/20cece67b5815f6e37ee72b86f282f027bab820b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L619-L623 and https://github.com/kubernetes/kubernetes/blob/20cece67b5815f6e37ee72b86f282f027bab820b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L669-L673

Comment (caesarxuchao) [reply pkg/registry/core/namespace/storage/storage.go]: Also, could you double check how the namespace controller will react if the namespace deletion is blocked by others?

The old code doesn't honor the foregroundDeletion, which could cause the namespace deletion to be blocked.

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 1 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: Sounds good.

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 3 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: Yes, new tests added.

File: pkg/registry/core/namespace/storage/storage.go (hunk 2 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage.go]: Ok, thanks for catching this. Now the behavior is changed to preserve finalizers if the delete options does not say explicitly.

> could you double check how the namespace controller will react if the namespace deletion is blocked by others?

Checking now.

Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage.go]: @caesarxuchao: Namespace controller does not care about finalizers in object meta too much, it only cares about the finalizer `kubernetes` in the spec.

If `kubernetes` finalizer is removed, and it gets an event, it just send a delete call again.

File: pkg/registry/core/namespace/storage/storage.go [hunk 4]
```diff
@@ -185,16 +185,38 @@ func (r *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOp
...
+				var orphanFinalizerAdded, deleteFinalizerAdded bool
+				// add orphan finalizer if we need to orphan this namespace.
+				if (options.OrphanDependents != nil && *options.OrphanDependents) ||
+					(options.PropagationPolicy != nil && *options.PropagationPolicy == metav1.DeletePropagationOrphan) ||
+					(options.OrphanDependents == nil && options.PropagationPolicy == nil && orphanFinalizerRemoved) {
+					newFinalizers = append(newFinalizers, metav1.FinalizerOrphanDependents)
+					orphanFinalizerAdded = true
+				}
+
+				// add foregroundDeletion if the options says foreground deletion.
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage.go]: ditto.

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 5]
```diff
@@ -189,6 +191,143 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
+			deleteOptions:       nil,
+			existingFinalizers:  []string{},
+			remainingFinalizers: map[string]bool{},
+		},
+		{
+			name: "propagation-background",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationBackground,
+			},
+			existingFinalizers: []string{
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: This should never happen, see https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L205.

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 6]
```diff
@@ -189,6 +191,143 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
+			},
+			existingFinalizers:  []string{},
+			remainingFinalizers: map[string]bool{},
+		},
+		{
+			name: "propagation-foreground",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationForeground,
+			},
+			existingFinalizers: []string{
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: Same here.

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 7]
```diff
@@ -189,6 +191,143 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerDeleteDependents: true,
+			},
+		},
+		{
+			name: "propagation-orphan",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationOrphan,
+			},
+			existingFinalizers: []string{
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: ditto

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 8]
```diff
@@ -189,6 +191,143 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
 }
 
+func TestDeleteWithGCFinalizers(t *testing.T) {
+	storage, server := newStorage(t)
+	defer server.Terminate(t)
+	defer storage.store.DestroyFunc()
+
+	propagationBackground := metav1.DeletePropagationBackground
+	propagationForeground := metav1.DeletePropagationForeground
+	propagationOrphan := metav1.DeletePropagationOrphan
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: nit: these definitions don't seem to improve readability.

File: pkg/registry/core/namespace/storage/storage_test.go [hunk 9]
```diff
@@ -189,6 +191,143 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
...
+		{
+			name: "propagation-orphan-without-finalizers",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationOrphan,
+			},
+			existingFinalizers: []string{},
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerOrphanDependents: true,
+			},
+		},
```
Comment (caesarxuchao) [inline pkg/registry/core/namespace/storage/storage_test.go]: Can you add a case if deleteOptions.orphanDependents=true?

Also, could you add a case where existing finalizer is "orphan", and the deleteOptions requires "foreground"?

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 8 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: `metav1.DeletePropagationBackground`(and the other twos) are constant, as a result, I can not take their address via `&metav1.DeletePropagationBackground`, I have to declare a new variable first.

File: pkg/registry/core/namespace/storage/storage.go (hunk 4 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage.go]: Ah, I must missed something, ditto to which comment?

Comment (caesarxuchao) [reply pkg/registry/core/namespace/storage/storage.go]: Please ignore it. I forgot to remove this comment.

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 8 above)
Comment (caesarxuchao) [reply pkg/registry/core/namespace/storage/storage_test.go]: Gotcha.

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 5 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: Done. (removed all the cases where existing finalizers contain both).

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 6 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: Done

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 7 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: Done

File: pkg/registry/core/namespace/storage/storage_test.go (hunk 9 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage_test.go]: Yeah, all added.

File: pkg/registry/core/namespace/storage/storage.go [hunk 10]
```diff
@@ -185,16 +185,38 @@ func (r *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOp
...
-				// Remove orphan finalizer if options.OrphanDependents = false.
-				if options.OrphanDependents != nil && *options.OrphanDependents == false {
-					// remove Orphan finalizer.
-					newFinalizers := []string{}
-					for i := range existingNamespace.ObjectMeta.Finalizers {
-						finalizer := existingNamespace.ObjectMeta.Finalizers[i]
-						if string(finalizer) != metav1.FinalizerOrphanDependents {
-							newFinalizers = append(newFinalizers, finalizer)
-						}
+				// remove orphan and foregroundDeletion first, add them back if needed
```
Comment (lavalamp) [inline pkg/registry/core/namespace/storage/storage.go]: Why is the logic structured this way? I am finding it very hard to follow. I would expect:

```
// 1. collect data
currentFinalizers := map[string]bool{}
for range existingNamespace.ObjectMeta.Finalizers { ... }

// 2. compute which finalizers it should have
shouldHaveFinalizers := map[string]bool{
  metav1.FinalizerOrphanDependents: shouldHaveOrphanFinalizer(options, ...),
  metav1.FinalizerDeleteDependents: shouldHaveForgroundFinalizer(options, ...),
}

// 3. do we need to make changes?
for k, v := range shouldHaveFinalizers {
  if v2, ok := currentFinalizers[k]; !ok || v != v2 {
    changeNeeded = true; break
  }
}

// 4. Make changes
if changeNeeded {
  ...
}
```

Putting the "shouldHaveXXXFinalizer" logic in a separate function will make it easier to unit test.

**Comment by lavalamp** [comment]
Why is the code different between namespace and other resources?

Thanks for the extensive tests.

File: pkg/registry/core/namespace/storage/storage.go (hunk 10 above)
Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage.go]: Ok, let me organize the code in the suggested way.

**Comment by zhouhaibing089** [comment]
> Why is the code different between namespace and other resources?

@lavalamp: It should not be different. And that's exactly the thing this change would like to fix.

File: pkg/registry/core/namespace/storage/storage.go (hunk 10 above)
Comment (lavalamp) [reply pkg/registry/core/namespace/storage/storage.go]: Thanks, sorry for being a pain!

Comment (zhouhaibing089) [reply pkg/registry/core/namespace/storage/storage.go]: Updated. Thanks! (That's a great suggestion!)

**Comment by zhouhaibing089** [comment]
Kindly ping @caesarxuchao and @lavalamp

File: pkg/registry/core/namespace/storage/storage.go (hunk 10 above)
Comment (lavalamp) [reply pkg/registry/core/namespace/storage/storage.go]: Thanks, much easier to follow now. (Next time we change this I'll recommend not mutating the "current" map, but this is fine for now. :) )

**Comment by caesarxuchao** [comment]
Note that the behavior of the namespace regarding finalizers is still slightly different from other built-in resources. When an UPDATE request removes the last metadata.finalizers, even if the spec.finalizers is empty, the namespace is not deleted by the apiserver. Clients need to send yet another DELETE request to delete the namespace.

CRD has the similar behavior.
